### ISSUE TITLE
Take advantage of "zero-copy" techniques from zmq blog

### ISSUE THREAD

**Issue by rotty** [erickt/rust-zmq#139]
See <http://zeromq.org/blog:zero-copy>. This may also make the `Sendable` trait and wrapping `zmq_send_const` (see #119) unnecessary, as it should be possible to provide more efficient implementations of `Into<Message>` for `&'static [u8]`. Even without specialization, we could provide an efficient implementation of `From<Vec<u8>>` for `Message`, i.e. one that moves data out of the vector and hands ownership over to zmq.

Before implementing this, it would be nice to add a basic micro-benchmark first as a baseline. Even better would be to transcribe a (hopefully) existing micro-benchmark employing "zero-copy", so Rust performance can be compared to C before and after implementing the optimization.

**Comment by birkenfeld** [comment]
Zero-copy is nice, but we have to consider the implications of callbacks from C - e.g. when the callback panics.

**Comment by rotty** [comment]
Yes, this a hairy subject to be approached with care and careful study of the nomicon.

**Comment by Kobzol** [comment]
> Even without specialization, we could provide an efficient implementation of `From<Vec<u8>>` for `Message`, i.e. one that moves data out of the vector and hands ownership over to zmq.

This one would be really nice to have, currently I can't find any way how to serialize a Protobuf object into a ZeroMQ message without redundant copying.

**Comment by Kobzol** [comment]
> Even without specialization, we could provide an efficient implementation of `From<Vec<u8>>` for `Message`, i.e. one that moves data out of the vector and hands ownership over to zmq.

This would be really nice, for example for serializing Protobuf objects into ZeroMQ messages.

**Comment by paddor** [comment]
In case anyone's looking for an alternative: [OMQ.rs](https://github.com/paddor/omq.rs) uses zero-copy `Bytes` payloads for owned/static data, inline storage for tiny messages, and gather-write for large sends.
